src/peer_digest.cc

```diff
@@ -548,27 +528,25 @@ peerDigestSwapInCBlock(void *data, char *buf, ssize_t size)
             fetch->state = DIGEST_READ_MASK;
             return StoreDigestCBlockSize;
         } else {
-            peerDigestFetchAbort(fetch, buf, "invalid digest cblock");
+            peerDigestFetchAbort(fetch, "invalid digest cblock");
             return -1;
         }
     }
 
     /* need more data, do we have space? */
     if (size >= SM_PAGE_SIZE) {
-        peerDigestFetchAbort(fetch, buf, "digest cblock too big");
+        peerDigestFetchAbort(fetch, "digest cblock too big");
         return -1;
     }
 
     return 0;       /* We need more data */
 }
 
 int
-peerDigestSwapInMask(void *data, char *buf, ssize_t size)
+peerDigestSwapInMask(DigestFetchState * const fetch, const char * const buf, const ssize_t size)
 {
-    DigestFetchState *fetch = (DigestFetchState *)data;
-    PeerDigest *pd;
+    const auto pd = fetch->pd.get();
 
```

Comment:
```suggestion
    assert(pd);
```

Reply:
Ok. Note that there is also another case in peerDigestFetchReply(), where fetch->pd goes unchecked... Before our changes there was an assertion in peerDigestHandleReply() (peerDigestFetchReply() caller), but we removed that assertion.

Reply:
That peerDigestFetchReply() use case is fine, I think: We are allowed to dereference `fetch->pd` without checking right after a peerDigestFetchedEnough() call because we know that peerDigestFetchedEnough() will protect us.  This peerDigestSwapInMask() check is different because we dereference `fetch->pd` before calling peerDigestFetchedEnough().

Needless to say, the whole thing is a big mess. We can probably simplify by checking for `pd` validity at the beginning of each fetching callback, never destroying `pd` in any fetching callbacks, and never checking `pd` anywhere else in fetching callbacks. However, that simplification spills into splitting peerDigestFetchedEnough() from "a universal check that we do not really know when to call because our state always changes as we make progress through callback code" to a set of change-specific checks. For now, let's not make those changes. We will revisit this when we know how the minimal fix looks like.

Reply:
Done (873e62).